src/scheduler/controller.rs

```diff
@@ -74,15 +74,20 @@ impl<VM: VMBinding> GCController<VM> {
         match message {
             CoordinatorMessage::Work(mut work) => {
                 work.do_work_with_stat(worker, mmtk);
-                self.scheduler
-                    .pending_messages
+                let old_count = self
+                    .scheduler
+                    .pending_coordinator_packets
                     .fetch_sub(1, Ordering::SeqCst);
+                if old_count == 1 {
+                    // When the coordinator finishes executing all coordinator work packets,
+                    // it is a chance to open more work buckets.
+                    // Notify one worker so it can open buckets.
+                    let _guard = self.scheduler.worker_monitor.0.lock().unwrap();
+                    self.scheduler.worker_monitor.1.notify_one();
```

Comment:
What confuses me is that although there is a queue for all the pending messages, currently we only send one work packet to the coordinator to stop all the mutators. For this stop mutators packet, it will call `notify_mutators_paused` to wake up all the workers [1]. So in theory there should be no deadlock issue?

Waking up workers here is a more general solution. Does that mean we can remove the wake-up code here [1]?

[1] https://github.com/mmtk/mmtk-core/blob/master/src/scheduler/scheduler.rs#L445

Reply:
`notify_mutators_paused` will wake up all workers **so that workers can scan mutator stacks**.  However, after scanning all stacks, the GC workers will sleep again.  Once all root-scanning work packets are done, and all GC workers are asleep, there will be nothing to wake up the worker **again**.

The lines I added here are for waking up the workers *the second time*, i.e. after all roots are scanned and all workers are sleeping because there are pending coordinator work packets.

This and `notify_mutator_paused` are independent, and cannot replace each other.  The code [1] represents one necessary condition (i.e. all mutators have stopped) to open "the first STW bucket" (i.e. Prepare).  The code I added fixes a problem that makes it fail to open the *second* STW bucket (i.e. Closure).

Reply:
I see